pkg/cluster/cluster.go

```diff
@@ -1108,3 +1114,49 @@ func (c *Controller) waitForHealthCheckAfterCreate(ctx context.Context, cluster
 	}
 	return nil
 }
+
+// maybeFixKubeConfigInsideContainer modifies the kubeconfig to set the KIND
+// apiserver's name and location to the name advertised on the `kind` docker
+// network (usually "$name-control-plane" where $name has the `kind-` prefix
+// removed), port 6443. This happens when the following conditions are true:
+//
+//   - The cluster to create is a KIND cluster
+//   - The effective DOCKER_HOST is `/var/run/docker.sock`
+//   - The environment variable $CTLPTL_IN_CONTAINER is set to a truthy value
+func (c *Controller) maybeFixKubeConfigInsideContainer(ctx context.Context, cluster *api.Cluster) error {
+	if cluster.Product != string(clusterid.ProductKIND) {
+		return nil
+	}
+
+	if c.dockerClient.DaemonHost() != "unix:///var/run/docker.sock" {
+		return nil
+	}
+
+	if inContainer, err := strconv.ParseBool(
+		strings.TrimSpace(os.Getenv("CTLPTL_IN_CONTAINER")),
```

Comment:
i think this is fine for now, but i do wonder if there's a better way to detect if we're running under the same docker instance we're connected to.

Reply:
maybe we could check if /etc/hostname looks like a container id, and ContainerInspect() says that container id is running?